making dictionary learning closer to the SparseNet algorithm

laurentperrinet opened the pull request:
The dictionary learning algorithm was assuming that the norm of the filters was equal to one. By using a heuristic to control for the norm of the filters, we allow for a more equilibrated learning. The implementation is a simplification of the one used in the original paper from Olshausen.

The dictionary learning is tested in:
https://laurentperrinet.github.io/sciblog/posts/2015-05-05-reproducing-olshausens-classical-sparsenet.html
and this PR is tested in:
https://laurentperrinet.github.io/sciblog/posts/2015-05-06-reproducing-olshausens-classical-sparsenet-part-2.html

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm \n The dictionary learning algorithm was assuming that the norm of the filters was equal to one. By using a heuristic to control for the norm of the filters, we allow for a more equilibrated learning. The implementation is a simplification of the one used in the original paper from Olshausen.

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm

GaelVaroquaux commented:
Thanks for the contribution.

The algorithm and problem formulation that we use is a fairly standard one (albeit different from the historical one of Olshausen). What is the practical benefit of this formulation?

Also, shouldn't the documentation be changed? The meaning of the gain parameter as well as its impact on the results should be explained in the narrative documentation.

Finally, isn't this a change in behavior? If it's the case, we need to have an option that maintains the old behavior, and a didactic example that shows the difference between the settings.

meduz commented:
Thanks for the comments!

First off, I made sure that by default there is no change to the results if the ``gain_rate`` setting is not used or set to zero. As such, there is no change in the behavior of the examples that I took from the sklearn website.

Now, you are  right that this ``gain``  vector certainly needs some clarification. I will try to update the ``_update_gain`` function to explicit that.

GaelVaroquaux commented:
> First off, I made sure that by default there is no change to the
> results if the gain_rate setting is not used or set to zero. As such,
> there is no change in the behavior of the examples that I took from the
> sklearn website.

That's great! I wasn't sure.

> Now, you are right that this gain vector certainly needs some
> clarification. I will try to update the _update_gain function to
> explicit that.

Great. Any chance that you could also make a small example illustrating
the impact of the gain and link it to the narrative documentation?

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm

meduz commented:
OK, thanks for the answer. I will try to make a minimal example of how it works.

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm : documenting the _update_gain function /  Mairal 2009 uses norm inferior or equal to 1

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm : Mairal 2009 uses norm inferior or equal to 1

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm : cleanly separating gradient learning (not yet implemented) and Dictionary learning from Mairal 2009

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm : cleanly separating gradient learning (implementing...) and Dictionary learning from Mairal 2009

laurentperrinet committed:
minor cosmetic changes to follow the work of Mairal, 2009

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm : cleanly separating gradient learning (implemented and testing) and Dictionary learning from Mairal 2009

laurentperrinet committed:
Merge branch 'mairal' of https://github.com/meduz/scikit-learn into sparsenet

* 'mairal' of https://github.com/meduz/scikit-learn:
  minor cosmetic changes to follow the work of Mairal, 2009

laurentperrinet committed:
making dictionary learning closer to the SparseNet algorithm : following the implementation from Olshausen

laurentperrinet committed:
tuning the homeostasis for SparseNet

laurentperrinet committed:
minor fixes before merging the code

laurentperrinet committed:
Merge branch 'master' into sparsenet

* master: (845 commits)
  MAINT: use xrange from six
  Update my website on whats_new.rst
  colorbar doesn't handle label parameter
  DOC Add description for UndefinedMetricWarning
  fontsize legend parameter is not available for matplotlib < 1.3
  FIX: gaussian_process will be deprecated in 0.19, not 0.18
  doc: fix typo
  FIX/ENH Make the moved class resilient to the deprecation patching
  MAINT move deprecated into deprecation.py
  MAINT Deprecated the UndefinedMetricWarning at sklearn.metrics.base
  DOC/SPHINX Add template for classes which don't have init (Exception classes)
  MAINT move custom error/warning classes into sklearn.exceptions
  ENH add pomegranate to related
  DOC: added entry for new GP implementation
  TST test_lml_precomputed() checks only for equality in first 7 digits
  ADD Mixins for normalized and stationary kernels
  FIX Enforcing y to be numeric in GPR and fixing python3 issue with maps
  ENH GPR.log_marginal_likelihood() returns the current log-likelihood if no theta vector is provided
  FIX String comparison via equality and not identity in Hyperparameter.__init__
  MISC Changing default for copy_X_train to True in GaussianProcessClassifier
  ...

laurentperrinet committed:
cleaning-up changes

meduz commented:
I have cleaned-up the changes to just focus on the replication of Olshausen's SparseNet algorithm. An example script is available @ https://github.com/meduz/SHL_scripts/blob/master/src/shl_scripts.py and the result may be seen @ https://github.com/meduz/SHL_scripts/blob/67a3286ba8b62a4390dd976b4fa305cadb9228d2/index.ipynb

amueller commented:
why did you add another class? And could you add an example (to the examples folder) showing the difference to the current version?

laurentperrinet committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into sparsenet

* 'master' of https://github.com/scikit-learn/scikit-learn: (283 commits)
  update file header in affinity_propagation.py
  DOC: Fix spelling
  adding 0.17 support link
  adding missing 0.16 doc link and removing broken support links
  Fixed unit test in sklearn/tree/tests/test_tree.py.
  fix link to contributing in website drop-down
  Fix Kernel PCA docstring to reflect how remove_zero_eig defaults to false
  removed the doc line 'Only applies if analyzer == 'word'' from CountVectorizer analyzer'
  Parallelize embarrassingly parallel loop in RFECV.fit
  Changed orders of docstring descriptions for polynomial_kernel
  Added more gamma info for sigmoid, polynomial kernels
  Added default gamma info for laplacian_kernel and rbf_kernel
  Add link to multiisotonic
  ENH add multinomial SAG solver for LogisticRegression
  change comment, printed comment, and code behavior to all reflect computation of mean squared error
  residual sum of squares should be computed by summing the squared differences instead of taking the mean.
  repaired dead link
  It's better to fix a comment
  Fix coveralls report sending
  edited docstring for clarity regarding the naming of the pipeline components.
  ...

laurentperrinet committed:
adding an example script to see the improvement in dictionary learning

laurentperrinet committed:
adding an example script to see the improvement in dictionary learning

meduz commented:
Thanks for your answer, this helps improving the PR.
- I have added a class to follow what was being done in the ``dict_learning.py`` file without interfering with existing methods. 
- I have now added an example script to show the 2 different set of filters with the existing method and that of Olshausen. To load an image I used ``imageio``, I hope this fits with guidelines within sklearn.

Cheers,
Laurent

amueller commented:
can you please paste the image here?
Unfortunately we can't use any non-standard libraries except the dependencies (numpy, scipy, matplotlib in the examples).

laurentperrinet committed:
adding an example script to see the improvement in dictionary learning : using only standard library

meduz commented:
I transformed the script to only use the standard library - should be better now. 

here are the results without and with the homeostasis used by Olshausen compared to the one used so far in sklearn:

![png image 516 x 567 pixels](https://cloud.githubusercontent.com/assets/381808/11815475/3142a614-a34c-11e5-8991-0310e245222d.png) 
![homeo png image 516 x 567 pixels](https://cloud.githubusercontent.com/assets/381808/11815482/3584aa4c-a34c-11e5-92f1-21063f991b33.png)

podshumok commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -545,6 +545,230 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+    ii = iter_offset - 1
+
+    for ii, this_X in zip(range(iter_offset, iter_offset + n_iter), batches):
+        dt = (time.time() - t0)
```
maybe you meant `timeit.default_timer()`?

meduz replied:
this bit of code is not related to this PR and is repeated for instance in ``dict_learning``, ``dict_learning_online `` and ``dict_learning_grad``

laurentperrinet committed:
adding some documentation in the dictionary learning script + small fix

meduz commented:
Any news on this PR? Thanks!

jnothman commented on examples/decomposition/plot_sparsenet.py:
```diff
@@ -0,0 +1,135 @@
+
+This example compares the set of filters obtained by sparse coding 
+algorithm described by Olshausen and Field in Nature, vol. 381,
+pp. 607-609 with that obtained using :ref:`DictionaryLearning`. The
```
I think you want :class:`sklearn.decomposition.DictionaryLearning`

laurentperrinet replied:
I based my writing on the `plot_image_denoising.py` script which is quite similar.

jnothman commented on examples/decomposition/plot_sparsenet.py:
```diff
@@ -0,0 +1,135 @@
+without homeostasis leads to the emergence of edge-like filters but that a 
+number of filters are not learned: They still are closer to noise and are 
+unlikely to be selected by the sparse coding algorithm without homeostasy.
+The heuristics used here follows the assumption that during learning,
```
follows -> follow; or heuristics -> heuristic

laurentperrinet replied:
done

jnothman reviewed (commented):
I hope to look at more of this soon, though it is outside of my expertise.

laurentperrinet committed:
Merge pull request #1 from scikit-learn/master

merging the master branch of sklearn into my branch

laurentperrinet committed:
gitignore

laurentperrinet committed:
typos

jnothman commented on .gitignore:
```diff
@@ -62,3 +62,13 @@ benchmarks/bench_covertype_data/
 *.sln
 *.pyproj
+
+sklearn/sklearn.egg-info/top_level.txt
```
No one should be building within the sklearn directory. Please revert these changes.

laurentperrinet replied:
Ok, I have reverted this change to the ``.gitignore`` file

jnothman reviewed (commented):
I've not looked at the code, but there is certainly an appreciable difference between the images in your plot above.

Please mention this to the narrative doc in `doc/modules/decomposition.rst`. Also add the new public API to `doc/modules/classes.rst`

jnothman commented:
You have a logic failure coming out in the tests when fitting:
```pytb
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/testing.py", line 741, in __call__
    return self.check(*args, **kwargs)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/testing.py", line 292, in wrapper
    return fn(*args, **kwargs)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/estimator_checks.py", line 460, in check_dict_unchanged
    estimator.fit(X, y)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/decomposition/dict_learning.py", line 1482, in fit
    return_n_iter=True
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/decomposition/dict_learning.py", line 739, in dict_learning_grad
    residual = this_X - np.dot(this_code.T, dictionary.T)
ValueError: matrices are not aligned
```

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -1081,6 +1307,184 @@ def fit(self, X, y=None):
+            Returns the instance itself.
+        """
+        random_state = check_random_state(self.random_state)
+        X = check_array(X)
```
you should perhaps use ensure_min_samples here, as the code is currently failing for a single-sample input.

laurentperrinet committed:
Merge branch 'master' into sparsenet

laurentperrinet committed:
using at least 2 samples

adding an option in ``check_array`` to have a minimum number of samples

jnothman commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-42d681bd97b726c5f8e4a5d822b2e5c0d3467695)

**new alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

laurentperrinet committed:
Fixed alert and documentation

Changed the example script to avoid alert and fixed reference to the
class.

laurentperrinet committed:
Fixed alert and documentation

Fixed reference to the class.

laurentperrinet committed:
Minimal documentation

Mentioning this to the narrative doc in doc/modules/decomposition.rst. Also adding to the new
public API in doc/modules/classes.rst

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -557,6 +557,232 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+    random_state = check_random_state(random_state)
+
+    if n_jobs == -1:
+        n_jobs = cpu_count()
```
I don't think you need to do this as it will be validated downstream

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -557,6 +557,232 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+    norm = np.sqrt(np.sum(dictionary**2, axis=1)).T
+
+    if verbose == 1:
+        print('[dict_learning]', end=' ')
```
This has no test coverage

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -557,6 +557,232 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+        X_train = X.copy()
+        random_state.shuffle(X_train)
+    else:
+        X_train = X
```
This has no test coverage

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -557,6 +557,232 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+        # Maybe we need a stopping criteria based on the amount of
+        # modification in the dictionary
+        if callback is not None:
+            callback(locals())
```
This has no test coverage.

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -557,6 +557,232 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+            print('Learning code...', end=' ')
+        elif verbose == 1:
+            print('|', end=' ')
+        code = sparse_encode(X, dictionary.T, algorithm=method, alpha=alpha,
```
This has no test coverage

jnothman commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -557,6 +557,232 @@ def dict_learning(X, n_components, alpha, max_iter=100, tol=1e-8,
+    if return_n_iter:
+        return dictionary.T, ii - iter_offset + 1
+    else:
+        return dictionary.T
```
This has no test coverage

jnothman reviewed (commented):
A few things here need testing (and there are some lines longer than 79 chars)

jnothman reviewed (commented):
A few things here need testing (and there are some lines longer than 79 chars)
